provider_gsuite.go

```diff
@@ -0,0 +1,217 @@
+func (g *GSuiteProvider) search(visited map[string]bool, userName string, depth int) error {
+	var newGroups []string
+	call := g.adminSvc.Groups.List().UserKey(userName).Fields("nextPageToken", "groups(email)")
+	if err := call.Pages(g.ctx, func(groups *admin.Groups) error {
+		for _, group := range groups.Groups {
+			if _, ok := visited[group.Email]; ok {
+				continue
+			}
+			visited[group.Email] = true
+			newGroups = append(newGroups, group.Email)
+		}
+		// Only recursively search for new groups that haven't been seen
+		if depth > 0 {
+			for _, email := range newGroups {
+				if err := g.search(visited, email, depth-1); err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// FetchUserInfo fetches additional user information from G Suite using custom schemas.
+func (g *GSuiteProvider) FetchUserInfo(b *jwtAuthBackend, allClaims map[string]interface{}, role *jwtRole) error {
+	if !g.config.FetchUserInfo || g.config.UserCustomSchemas == "" {
+		return nil
+	}
+
+	userName, err := g.getUserClaim(b, allClaims, role)
+	if err != nil {
+		return err
+	}
+
+	// Set context and create a new admin service for requests to Google admin APIs
+	g.ctx = b.providerCtx
+	g.adminSvc, err = admin.NewService(g.ctx, option.WithHTTPClient(g.jwtConfig.Client(g.ctx)))
+	if err != nil {
+		return err
+	}
+
+	return g.fillCustomSchemas(userName, allClaims)
+}
+
+// fillCustomSchemas fetches G Suite user information associated with the custom schemas
+// configured for this provider. It inserts the schema -> value pairs into the passed
+// allClaims so that the values can be used for claim mapping to token and identity metadata.
+func (g *GSuiteProvider) fillCustomSchemas(userName string, allClaims map[string]interface{}) error {
+	userResponse, err := g.adminSvc.Users.Get(userName).Context(g.ctx).Projection("custom").
+		CustomFieldMask(g.config.UserCustomSchemas).Fields("customSchemas").Do()
+	if err != nil {
+		return err
+	}
+
+	for schema, rawValue := range userResponse.CustomSchemas {
+		// note: metadata extraction via claim_mappings only supports strings
+		// as values, but filtering happens later so we must use interface{}
+		var value map[string]interface{}
+		if err := json.Unmarshal(rawValue, &value); err != nil {
+			return err
+		}
+
+		allClaims[schema] = value
+	}
+
+	return nil
+}
+
+// getUserClaim returns the user claim value configured in the passed role.
+// If the user claim is not found or is not a string, an error is returned.
+func (g *GSuiteProvider) getUserClaim(b *jwtAuthBackend, allClaims map[string]interface{}, role *jwtRole) (string, error) {
+	userClaimRaw := getClaim(b.Logger(), allClaims, role.UserClaim)
+	if userClaimRaw == nil {
+		return "", fmt.Errorf("unable to locate '%s' in claims", role.UserClaim)
+	}
+	userClaim, ok := userClaimRaw.(string)
+	if !ok {
+		return "", fmt.Errorf("claim '%s' could not be converted to string", role.UserClaim)
```

Comment:
```suggestion
		return "", fmt.Errorf("claim %q could not be converted to string", role.UserClaim)
```

Reply:
Done in 34fa56b8fa81a5e27a42e4b58e6d7ffeabbdc7cb.